### PR TITLE
Feature/domain queue join

### LINKED ISSUE #3561: v9-gui: Add the ability to join rejoin and unjoin domains via the api

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
Allow to start a pfqueue task asynchronously and poll its status
Use the above for long-running domain operations

# Impacts
pfqueue plumbing
Unified API

# Issue
fixes #3561 

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

**Comment by julsemaan** [comment]
Example:
https://support.inverse.ca/~jsemaan/domain-join.html

File: lib/pf/pfqueue/status_updater/redis.pm [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+sub set_in_status_hash {
+    my ($self, $key, $data) = @_;
+    if($self->finalized) {
+        get_logger->trace("Not reporting data for $key since status has been finalized");
+        return;
+    }
+    
+    $self->connection->multi();
+    $self->connection->expire($self->status_key, $self->status_ttl);
```
Comment (jrouzierinverse) [inline lib/pf/pfqueue/status_updater/redis.pm]: The expire should be done after the hset.
Because if the hash does not exists the expire will fail.

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]
Redis PUB/SUB is unreliable

The current approach will cause some messages to be lost.
If the subscriber is not listening when the message is published that message will be lost.

The only way to avoid losing message is to use BRPOPLPUSH/RPOPLPUSH.

File: go/caddy/job-status/job-status.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"git.inverse.ca/inverse/fingerbank-processor/sharedutils"
+
+	"github.com/go-redis/redis"
```
Comment (jrouzierinverse) [inline go/caddy/job-status/job-status.go:10]: We already have a redis connection pool
"github.com/inverse-inc/packetfence/go/redisclient"

This uses the config of pfqueue to connect to redis.
I know it is too late to rework at least use the config to connect to redis.

**Comment by julsemaan** [comment]
We don't really care if a publication gets lost since it doesn't contain anything important as the status key holds all the data.

Receiving a publication on the channel just allows the polling to complete faster. The worst that can happen is that the user will wait 15 seconds because of a missed publication that occurred between his 2 requests. The result of the operation will be the same though.

To me that isn't a big issue but if that's a blocker to the team, I can adjust it but the publisher (pfqueue) shouldn't be blocking and ensuring his publication went through and was received by someone as there is no guarantee someone will be polling the status. Someone could just give up on polling the status of the job or some jobs could be started with intent of polling but the consumer just decides to let it do its job without polling.

**Comment by julsemaan** [comment]
As verbally discussed, I made polling exit early if the job has already completed

File: go/caddy/job-status/job-status.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+	return jobStatus, nil
+}
+
+func (h JobStatusHandler) handleStatusPoll(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+	ctx := r.Context()
+	jobId := p.ByName("job_id")
+	updatesKey := h.jobStatusUpdatesKey(jobId)
+
+	sub := h.redis.Subscribe(updatesKey)
```
Comment (jrouzierinverse) [inline go/caddy/job-status/job-status.go:106]: defer sub.Close() to avoid a resource leak

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

**Comment by julsemaan** [comment]
Performed the requested adjustments

**Comment by jrouzierinverse** [comment]
rebase

**Comment by julsemaan** [comment]
Adjustments performed for password handling + rebased

**Comment by jrouzierinverse** [comment]
rebase then I will merge

**Comment by julsemaan** [comment]
rebased

**Pull request merged**